storage/innobase/row/row0import.cc

```diff
@@ -4246,6 +4442,173 @@ fil_tablespace_iterate(
+{
+  dberr_t err= DB_SUCCESS;
+  pars_info_t*    info = nullptr;
+  for (ulint n= table->n_cols - 1; n >= fts_pos; n--)
+  {
+    info= pars_info_create();
+    pars_info_add_ull_literal(info, "id", table->id);
+    pars_info_add_ull_literal(info, "pos", n);
+    pars_info_add_ull_literal(info, "val", n + 1);
+    err= que_eval_sql(info,
+                    "PROCEDURE UPD_ALL_COL () IS\n"
+                    "BEGIN\n"
+                    "UPDATE SYS_COLUMNS SET\n"
+                    "POS = :val\n"
+                    "WHERE TABLE_ID = :id AND POS = :pos;\n"
+                    "END;\n", trx);
+    if (err != DB_SUCCESS)
+    {
+      my_error(ER_INTERNAL_ERROR, MYF(0),
+               "InnoDB: Updating SYS_COLUMNS failed");
+      return err;
+    }
+  }
+
+  info= pars_info_create();
+  dict_col_t* fts_col= &table->cols[fts_pos];
+  pars_info_add_ull_literal(info, "id", table->id);
+  pars_info_add_int4_literal(info, "pos", fts_pos);
+  pars_info_add_str_literal(info, "name", FTS_DOC_ID_COL_NAME);
+  pars_info_add_int4_literal(info, "mtype", fts_col->mtype);
+  pars_info_add_int4_literal(info, "prtype", fts_col->prtype);
+  pars_info_add_int4_literal(info, "len", fts_col->len);
+  pars_info_add_int4_literal(info, "base", 0);
+
+  err= que_eval_sql(info,
+                    "PROCEDURE ADD_COL () IS\n"
+                    "BEGIN\n"
+                    "INSERT INTO SYS_COLUMNS VALUES"
+                    "(:id,:pos,:name,:mtype,:prtype,:len,:base);\n"
+                    "END;\n", trx);
+  if (err != DB_SUCCESS)
+  {
+    my_error(ER_INTERNAL_ERROR, MYF(0),
+             "InnoDB: Insert into SYS_COLUMNS failed");
+    return err;
+  }
+
+  info= pars_info_create();
+  ulint n= dict_table_encode_n_col(
+    unsigned(table->n_cols) - DATA_N_SYS_COLS, table->n_v_cols)
+    | (table->flags & DICT_TF_COMPACT) << 31;
+
+  pars_info_add_int4_literal(info, "n", n);
+  pars_info_add_ull_literal(info, "id", table->id);
+
+  err= que_eval_sql(info,
+                    "PROCEDURE UPDATE_N_COLS () IS\n"
+                    "BEGIN\n"
+                    "UPDATE SYS_TABLES SET N_COLS = :n"
+                    " WHERE ID = :id;\n"
+                    "END;\n", trx);
+  if (err != DB_SUCCESS)
+    my_error(ER_INTERNAL_ERROR, MYF(0),
+             "InnoDB: Updating SYS_TABLES.N_COLS failed");
+  return err;
+}
+
+/* Need to check whether the table need to add system generated
+fts column and system generated fts document index
+@param cfg   config file
+@param table table to be imported
+@return whether the table has to add system generated fts column
+and fts index */
+bool need_hidden_fts(row_import& cfg, dict_table_t *table)
+{
+  if (cfg.m_missing || !cfg.has_hidden_fts()) return false;
+
+  ulint col_name_len= 0;
+  for (ulint i= 0; i < table->n_cols; i++)
+  {
+     if (!strcmp(&table->col_names[col_name_len],
+                 FTS_DOC_ID_COL_NAME))
+       return false;
+     col_name_len += strlen(&table->col_names[col_name_len]) + 1;
```

Comment:
We could replace `col_name_len` with a straight pointer.
```c++
  const char *col_name= table->col_names;
  for (unsigned i= 0; i < table->n_cols; i++, col_name+= strlen(col_name) + 1)
    if (!strcmp(col_name, FTS_DOC_ID_COL_NAME)
       return false;
```
But, better yet, maybe we should just replace the entire function with the following:
```c++
return !cfg.m_missing && cfg.has_hidden_fts() && !table->fts_doc_id_index;
```
The `m_missing` check might better be part of `bool row_import::has_hidden_fts() const`.